File: python/querygraph/did.py
from __future__ import annotations

from dataclasses import dataclass
from hashlib import sha256

from querygraph.base58 import b58encode


@dataclass(frozen=True)
class DidDocument:
    id: str
    controller: str
    public_key_multibase: str
    context: list[str] | None = None
    service_endpoint: str | None = None

    @classmethod
    def new_oyd(cls, seed: bytes | str, controller: str) -> "DidDocument":
        seed_bytes = seed.encode() if isinstance(seed, str) else seed
        digest = sha256(seed_bytes).digest()
        multihash = bytes([0x12, 0x20]) + digest
        fingerprint = b58encode(multihash)
        return cls(
            context=[
                "https://www.w3.org/ns/did/v1",
                "https://w3id.org/security/suites/ed25519-2020/v1",
            ],
            id=f"did:oyd:z{fingerprint}",
            controller=controller,
            public_key_multibase=f"z{b58encode(digest)}",
        )

    def with_service_endpoint(self, endpoint: str) -> "DidDocument":
        return DidDocument(
            context=self.context,
            id=self.id,
            controller=self.controller,
            public_key_multibase=self.public_key_multibase,
            service_endpoint=endpoint,
        )

    def to_json(self) -> dict:
        doc = {
            "id": self.id,
            "controller": self.controller,
            "public_key_multibase": self.public_key_multibase,
            "service_endpoint": self.service_endpoint,
        }
        if self.context is not None:
            doc["@context"] = self.context
        return doc

File: src/did.rs
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct DidDocument {
    #[serde(rename = "@context", skip_serializing_if = "Option::is_none")]
    pub context: Option<Vec<String>>,
    pub id: String,
    pub controller: String,
    pub public_key_multibase: String,
    pub service_endpoint: Option<String>,
}

impl DidDocument {
    pub fn new_oyd(seed: impl AsRef<[u8]>, controller: impl Into<String>) -> Self {
        let digest = Sha256::digest(seed.as_ref());
        let mut multihash = Vec::with_capacity(34);
        multihash.push(0x12);
        multihash.push(0x20);
        multihash.extend_from_slice(&digest);
        let fingerprint = bs58::encode(multihash).into_string();
        let id = format!("did:oyd:z{fingerprint}");
        Self {
            context: Some(vec![
                "https://www.w3.org/ns/did/v1".to_string(),
                "https://w3id.org/security/suites/ed25519-2020/v1".to_string(),
            ]),
            id,
            controller: controller.into(),
            public_key_multibase: format!("z{}", bs58::encode(digest).into_string()),
            service_endpoint: None,
        }
    }

    pub fn with_service_endpoint(mut self, endpoint: impl Into<String>) -> Self {
        self.service_endpoint = Some(endpoint.into());
        self
    }
}
